src/Controllers/ShareDraftController.php

```diff
@@ -115,6 +120,50 @@ public function preview(HTTPRequest $request)
         }
     }
 
+    /**
+     * With the rendered HTML, regex match files that are in the protected asset store and grant access
+     * if they have no view permissions on them.  This means they are only in the protected asset store
+     * because they are unpublished
+     *
+     * This will also correctly handle modified files (different version on draft vs live)
+     *
+     * Doing a regex on rendered HTML, rather than using shortcodes, because ShareDraftController has no
+     * underlying understanding of how HTML is generated (e.g. $page->Content vs content blocks vs other methods).
+     * Instead it does a "fake request" to a page url to get the rendered HTML
+     *
+     * This isn't ideal as it's making an assumption on the protected asset store url structure.
+     * Ideally we'd do this on shortcodes higher up around where `$page` is defined, though we'd need a
+     * "Content API" to handle things such as content blocks.
+     *
+     * @param string $rendered
+     */
+    private function grantAccessToDraftFiles(string $rendered): void
+    {
+        // Note: image in folder is /assets/myfolder/cdce3f8a1a/sample-kokako__ResizedImageWzYwMCw0NTBd.jpg
+        // Making the assumption that shortcode providers will always use double quotes for attributes
+        $regex = '#<(img|a)[^>]+(src|href)="([^"]*?)' . ASSETS_DIR . '/([^"]*?/?([a-z0-9]{10})/[^./]+\.[a-z0-9]+)"#';
+        preg_match_all($regex, $rendered, $matches);
+        $urls = $matches[4];
+        $hashes = $matches[5];
+        // Use NaturalFileIDHelper to clean up variants such as __Resized images
+        $helper = NaturalFileIDHelper::create();
```

Comment:
Can't really assume what Helper will be used here or what resolution strategy.

Reply:
Updated